### ISSUE TITLE
Offline wiki

### ISSUE THREAD

**Issue by lethosor** [DF-Wiki/DF-Wiki#3]
It looks like this was never migrated from http://dwarffortresswiki.org/index.php/User_talk:Briess#Cloning_the_wiki:

> 1. Massage the DF2014 namespace into a set of static HTML documents
> 2. Bundle these documents in the docset format
> 3. Release the docset in Dash on OSX and Zeal on Windows and Linux.

As discussed before, in order for this to work without impacting the servers, a clone of the wiki would need to be set up locally to generate HTML files (which requires parts of LocalSettings.php to be made public).

**Comment by lethosor** [comment]
@ramblurr Are you still interested in working with https://github.com/Ramblurr/Dwarf-Fortress-Portable-Wiki, or would you prefer to generate HTML documentation with a Mediawiki instance?

**Comment by Ramblurr** [comment]
@lethosor the latter I think. DF-Wiki uses some great templates that provide a lot of info and manually implementing those (which is what Ramblurr/Dwarf-Fortress-Portable-Wiki does) is just too much work.

IMHO, the best way to accomplish #1 is to get a mirror of df-wiki running locally, spider it (again locally) into a set of static documents that can be bundled together offline.

It could even be automated in the end, I just need a reliable way to make a mirror of DF-wiki (or just the DF201X namespace and images) that doesn't blow through your bandwidth.

**Comment by locriani** [comment]
Bandwidth isn't an issue, the issues we'll need to resolve are:
1. CPU time usage
2. Sane DFWiki mirror

We probably need to pull the sensitive config files out of the dfwiki-files repo anyways which could get us closer to the goal.

**Comment by lethosor** [comment]
@locriani Are you proposing a mirror on our end or making dfwikifiles public?

**Comment by locriani** [comment]
Just throwing out ideas more like.

**Comment by lethosor** [comment]
I think setting up something to generate a simpler dump format might be easier to work with (i.e. titles and contents only), and would solve issues like #2 as well. @Ramblurr thoughts?

**Comment by Ramblurr** [comment]
Its not so much about the format of the dump that is the problem (though, its true parsing wikimedia xml dumps is a pain), but that the dump's contents are raw wikitext before the template processing. 

A simple dump format that includes the post-processed titles and contents would work great. But then as @locriani mentioned, processing the wiki text is a cpu bound task.

**Comment by lethosor** [comment]
I was thinking of raw wikitext, actually. One of the problems with the current dump is that DF2014 pages' contents are listed in the mainspace, probably due to AutoRedirect, so I thought that we might as well simplify the dump format if we have to use something else. (Of course, there could be problems that occur with hundreds of files in a directory.)

I was under the impression that the size of an HTML dump was more of an obstacle than processing power. Rendering only the bodies of pages could cut this size in half, but it would still be larger than a wikitext dump.

**Comment by acroca** [comment]
Any updates here? I'm interested in getting the wiki offline too. I can help you guys if you need something too.

**Comment by locriani** [comment]
@acroca We're currently working on getting the wiki fully upgraded to 1.26, and resolving some outstanding mail issues before we tackle this again.

**Comment by acroca** [comment]
Ok no problem. In the meantime I've been trying to get the xml dump and import it in a new installation of mediawiki, but I'm not sure that'll be enough since the dump is pretty small and a lot of things won't be there.

**Comment by Ramblurr** [comment]
Yea, it's quite tricky to get the dump into a useful state. Most of the interesting data on DF Wiki is bound up in templates.

I wrote some scripts that iterated through the dump, then used the MW api to render the templates to HTML. But that requires an HTTP request per page to the live instance, which is no good at all.

The "easiest" solution would be to periodically (monthly?) duplicate the DF wiki on a virtual machine, then run a local crawler to crawl the rendered pages.

**Comment by gamma** [comment]
Are there any updates on the topic? The biggest problem is to have a fresh dump of the content and images. The rest could be done offsite (VM/Docker).

**Comment by lethosor** [comment]
I don't remember much happening with this. Images have come up before, and they'd probably make the dump too large, so I'm not sure about what to do there.

**Comment by Ramblurr** [comment]
Yea. To make it work well, the host of the wiki would need to provide periodic dumps and make them available, but that would incur a significant bandwidth cost.. not sure if the host is willing to do that.

**Comment by gamma** [comment]
Thanks for the comments. The last dump I saw was about 200MB. What I specifically would be interested in is a dump of the most recent version including its images. I have no idea how media wiki handles those.

I'd suggest a monthly or bi-monthly dump including images with a direct upload to github LFS or any other offsite storage (even a Docker-Data container would be great. Depending on the access to the host and what software is available. The data could be packaged and pushed to docker hub).

You could even overwrite the latest version This way you would have to handle with a lot of traffic for recurring downloads.

**Comment by Ramblurr** [comment]
A github LFS repo or docker container would be awesome.

**Comment by Hello71** [comment]
the standard method is to provide a raw wikitext dump in the Special:Export format. there are about 30k pages on http://dwarffortresswiki.org/ presently according to Special:Statistics. assuming arbitrarily that an average page will require 1 kB after gzip (which I obtained by pressing "Random page" then compressing that), each dump will require 30 MB of space.

files are a separate issue.

**Comment by locriani** [comment]
Except, the job that we used to have generate the dump took over 28 hours to complete, because of the way MediaWiki page processing works for exports. (It can't rely on the cache). If this is going to happen, it will need to happen offline to avoid a ridiculous spike in server costs.

**Comment by Hello71** [comment]
https://google.com/search?q=mediawiki+dump

**Comment by lethosor** [comment]
I believe dumpBackup.php is what was used before. I don't think fetching and rendering 30k pages is particularly fast, although I haven't looked into it enough to know why it was *that* slow.

**Comment by Hello71** [comment]
the whole point of dumpBackup is that it makes an XML dump. an HTML dump would be useless as a backup because you wouldn't be able to import it into another MW.

**Comment by gamma** [comment]
I'd be glad to have a regular XML dump - shure additional images in another zip would be awesome too.

**Comment by Ramblurr** [comment]
I already went down the XML dump path as described [above](https://github.com/DF-Wiki/DF-Wiki/issues/3#issuecomment-165983317).

> Yea, it's quite tricky to get the dump into a useful state. Most of the interesting data on DF Wiki is bound up in templates.
>
> I wrote some scripts that iterated through the dump, then used the MW api to render the templates to HTML. But that requires an HTTP request per page to the live instance, which is no good at all.

**Comment by Hello71** [comment]
my point is that with an XML dump you can install MW locally to render the pages, either statically or on-the-fly.

**Comment by Ramblurr** [comment]
Yes, but if you roll with vanilla MW and load up the DF wiki dump, it will not be that useful.

Not without:

1. The DF wiki theme / css / js etc
2. All the DF wiki plugins to render the templates
3. The images and other files which aren't included in the dump

I'm not saying we shouldn't have the dump, just cautioning that it will be of limited use if your goal is a *useable* offline version of the DF wiki.

**Comment by gamma** [comment]
This sure is a thing and I guess the images would be the biggest trouble. AFAIK at least some of the custom plugins are already available via Github?

Out of curiosity: you said 30k pages all together - is it possible to only dump pages and images et all from the most recent DF version?

**Comment by Hello71** [comment]
> 1. The DF wiki theme / css / js etc

the skin is standard Vector, afaict. customizations are in the MediaWiki namespace (constitutes part of "pages")

> 2. All the DF wiki plugins to render the templates

the only extensions not hosted by the MW project/WM are those written by lethosor, which all appear to be on GH. in general, MW extensions are trivial to install, and these seem no different (although somewhat poorly documented).

> 3. The images and other files which aren't included in the dump

yes, these are somewhat of a problem, but can be easily handled outside of the MW framework. the only reason why WM doesn't provide them is because of the large size. I can't query the size of files because your MW installation is far too old (4 years and counting).

**Comment by Hello71** [comment]
> Out of curiosity: you said 30k pages all together - is it possible to only dump pages and images et all from the most recent DF version?

yes, the pages are in a separate namespace. images must be dumped separately anyways.

**Comment by bobberb** [comment]
@lethosor what is the most up to date offline source given the recent downtime?

**Comment by lethosor** [comment]
I'm not sure. I would say the XML dump, but I can't quickly find a link to it. That's probably a better question for locriani.

**Comment by bobberb** [comment]
@locriani

**Comment by gamma** [comment]
I have a dump from about 6 months ago.

**Comment by Ramblurr** [comment]
@gamma does it have images?

Personally I think the best way to create a usable offline version is to setup a separate instance of the wiki (locally perhaps), then scrape it, using a tool like httrack, then munge the scraped output into some package that is easily used offline.

Why a separate instance? To prevent load and increased cost on the production server.

**Comment by gamma** [comment]
afaik I have the images too. The dump is quite big though. I used the wikiteams dumpgenerator tool for it.

**Comment by TurnOnMind** [comment]
Hi, if topic is still on top. I would have an idea to generate updated wiki file every time when someone download it. Without high server overload, and without user time consuming page by page loading.

**Comment by lethosor** [comment]
How would you do that?
